syntax/zsh.vim

```diff
@@ -140,7 +140,7 @@ syn keyword zshCommands         alias autoload bg bindkey break bye cap cd
                                 \ enable eval exec exit export false fc fg
                                 \ functions getcap getln getopts hash history
                                 \ jobs kill let limit log logout popd print
-                                \ printf pushd pushln pwd r read
+                                \ printf prompt pushd pushln pwd r read
```

Comment:
This is a list of builtin commands, which `prompt` is not; it's merely an autoloadable function shipped with zsh itself.  See:

```zsh
% type printf prompt pushd pushln pwd r read 
printf is a shell builtin
prompt not found
pushd is a shell builtin
pushln is a shell builtin
pwd is a shell builtin
r is a shell builtin
read is a shell builtin
zsh: exit 1
```

Reply:
@danielshahaf What `zrecompile` then?

And what about all these?
```zsh
zsh --no-rcs -c 'type zrecompile clone cap stat zftp zprof zpty zsocket ztcp'
```

Reply:
Most of these are from `zshbuiltins(1)`, instead of `zshcontrib(1)`.

Maybe we want to include both? As an aside, I think it would be better to generate this from the zsh source, similar to `make-options.zsh`; here's a punt:

	#!/bin/zsh

	(( ! $+argv[1] )) && print >&2 'set the first argument to a zsh source direcrory' && exit 1
	src=$1

	gen() {
		local file=$1
		local IFS=$'\n'
		typeset -U all=()

		for opt in $(grep '^findex([A-Za-z_]*)$' $file); do
			all+=(${${(L)opt#findex\(}%\)})
		done
		for opt in $(grep '^zlecmd([A-Za-z_]*)$' $file); do
			all+=(${${(L)opt#zlecmd\(}%\)})
		done
		for opt in $(grep '^alias([A-Za-z_]*)(.*)$' $file); do
			all+=(${${(L)opt#alias\(}%%\)*})
		done
		for opt in $(grep '^module([A-Za-z_]*)(.*)$' $file); do
			all+=(${${(L)opt#module\(}%%\)*})
		done
		for opt in $(grep '^item(tt([A-Za-z_]*)' $file); do
			all+=(${${(L)opt#item\(tt\(}%%\)*})
		done
		for opt in $(grep '^xitem(tt([A-Za-z_]*)' $file); do
			all+=(${${(L)opt#xitem\(tt\(}%%\)*})
		done

		all=(${(o)all})
		lines=($(fold -sw100 <<<${(j: :)all}))
		print -r 'syn keyword zshCommands'
		print    "           \\\ ${(j:\n           \\ :)lines}"
	}

	gen $src/Doc/Zsh/builtins.yo
	gen $src/Doc/Zsh/contrib.yo

Outputs:

    [~pack/vim-zsh](master)% ./make-commands.zsh ~/src/zsh/
    syn keyword zshCommands
               \ alias autoload bg bindkey break builtin bye cap cd chdir clone command comparguments compcall
               \ compctl compdescribe compfiles compgroups compquote comptags comptry compvalues continue cube
               \ declare dirs disable disown echo echotc echoti emulate enable eval exec exit export false fc fg
               \ float functions getcap getln getopts hash history integer jobs kill let limit local logout popd
               \ print printf pushd pushln pwd r read readonly rehash return sched set setcap setopt shift source
               \ stat suspend test times trap true ttyctl type typeset ulimit umask unalias unfunction unhash
               \ unlimit unset unsetopt vared wait whence where which zcompile zformat zftp zle zmath_cube zmodload
               \ zparseopts zprof zpty zregexparse zsocket zstyle ztcp
    syn keyword zshCommands
               \ _cdr actionformats bash bg bg_bold bg_no_bold branchformat catch cdr chpwd_recent_add
               \ chpwd_recent_dirs chpwd_recent_filehandler color colors colour command complete completer context
               \ cursor debug default disable disown enable fg fg_bold fg_no_bold files flags fned formats g handler
               \ hgrevformat histed hook_com hooks incarg key list mailcap match max min normal nslookup nvcsformats
               \ p pager parent place pop priority prompt prompt_ quiltcommand reporter ret rprompt running s shell
               \ specified stagedstr sum tetris tetriscurses throw toggle tty unspecified unstagedstr vcs vcs_info
               \ vcs_info_hookadd vcs_info_hookdel vcs_info_lastmsg vcs_info_printsys vcs_info_setsys verbose
               \ whitespace widget x zargs zcalc zcp zed zkbd zln zmathfunc zmathfuncdef zmv zrecompile
               \ zsh_directory_name_generic

Reply:
There's a few in that second `zshCommands` that are wrong: `bash`, `normal`, `specified`, etc. It (wrongly) matches a list of items in the ZLE FUNCTIONS section. Some commands are also in both lists. Need to look at it later. Pushed things over here for now: https://github.com/arp242/vim-zsh/tree/gen-commands

Reply:
Hey, I am happy to include whatever you come to a conclusion :) just ping me once this is resolved :)

Reply:
@arp242 I don't know, seems like a lot of cruft.

It's hard to view what is changing from the diff, so these would be the commands being added:

```
actionformats bg bg_bold bg_no_bold branchformat builtin catch _cdr cdr chpwd_recent_add chpwd_recent_dirs chpwd_recent_filehandler color colors colour command command complete completer context cube cursor debug declare disable disown enable fg fg_bold fg_no_bold files flags float fned formats g handler hgrevformat histed hook_com hooks incarg integer key list local mailcap match max min nslookup nvcsformats p pager parent place pop priority prompt_ quiltcommand readonly reporter ret rprompt running s setopt specified stagedstr sum tetris tetriscurses throw toggle tty typeset unsetopt unspecified unstagedstr vcs vcs_info vcs_info_hookadd vcs_info_hookdel vcs_info_lastmsg vcs_info_printsys vcs_info_setsys verbose widget x zargs zcalc zcp zed zkbd zln zmath_cube zmathfunc zmathfuncdef zmv zsh_directory_name_generic
```

Reply:
Yes, the contrib file matches too much; I wrote it for builtins.yo first, and then assumed (hoped) it would also work well for contrib.yo, but it doesn't really.

Reply:
And I see it has problems parsing prompt stuff.

This:

    item(tt(prompt_)var(theme)tt(_setup))

Is interpreted as `prompt_`.

This:

    Use the tt(prompt) command to select your preferred theme.

Isn't detected.

And these:

```
kindex(prompt, nslookup style)
kindex(rprompt, nslookup style)
xitem(tt(prompt))
item(tt(rprompt))
```

Get interpreted only as `rprompt`, but they are styles, not commands.

Reply:
@arp242 There's issues with `builtins.yo`: it matches an example and `zmath_cube` and `cube` are added. Moreover, some of the commands already in `zshPrecommand` or other groups. I don't think it can be automated that easily.

However, I took your script and generated a helper of my own:

```zsh
grep --color=always -C 3 \
	-e '^findex([A-Za-z_]*)$' \
	-e '^zlecmd([A-Za-z_]*)$' \
	-e '^alias([A-Za-z_]*)(.*)$' \
	-e '^module([A-Za-z_]*)(.*)$' \
	-e '^item(tt([A-Za-z_]*)' \
	-e '^xitem(tt([A-Za-z_]*)' \
	$1/Doc/Zsh/builtins.yo | less -R
```

With this I'm able to quickly see what might be good to include. I manually created a list and compared to what is already listed.

The only difference is `log` and `zrecompile`. `log` is not a builtin command, but part of the `zsh/watch` module. It probably should not be listed. That's basically it.

Also, `exec` is listed twice (in `zshCommands` and `zshPrecommand`).

Reply:
Yeah, it's not perfect :-/

I'm fine with just adding `prompt` for now; I don't have the time to pick this up right now, although I think splitting it out to `zshBuiltin` and `zshContrib` would be a good idea (Also for the other non-builtins in `zshCommands`). This way people can remove the highlights for it if they want, or change it.